policy_agent/policy_agent.py

```diff
@@ -0,0 +1,643 @@
+                             (kind, target_key))
+                self.process_resource(action="ADDED", kind=kind, target=target)
+
+        else:
+            _log.error("Event in process_resource not recognized: %s" % r_type)
+
+    def delete_resource(kind, target):
+        if kind == "Namespace":
+            resource_pool = self.namespaces
+            obj = Namespace(target)
+
+        elif kind == "Service":
+            resource_pool = self.services
+            obj = Service(target)
+
+        elif kind == "Pod":
+            resource_pool = self.pods
+            obj = Pod(target)
+
+        elif kind == "Endpoints":
+            resource_pool = self.endpoints
+            obj = Endpoints(target)
+
+        target_key = obj.key
+
+        if target_key in resource_pool:
+            del resource_pool[target_key]
+            _log.info("%s %s deleted from Calico store" % (kind, target_key))
+
+        else:
+            _log.error(
+                "Tried to Delete %s %s but it was not in bin" %
+                (kind, target_key))
+
+    def match_pod(self, namespace, name):
+        try:
+            return self.pods["%s/%s" % (namespace, name)]
+        except KeyError:
+            _log.error("Pod %s in NS %s not processed" % (name, namespace))
+            return None
+
+    def resync(self):
+        """
+        Updates Calico store with new resource information
+        """
+        # Create/update Namespace Profiles
+        for namespace_key, ns in self.changed_namespaces.items():
+            if ns.create_ns_profile():
+                self.nsRV = ns.RV
+                self.namespaces[namespace_key] = ns
+                del self.changed_namespaces[namespace_key]
+
+        # Process Services
+        for service_key, svc in self.changed_services.items():
+            self.svcRV = svc.RV
+            self.services[service_key] = svc
+            del self.changed_services[service_key]
+
+        # Apply Namespace policy to new/updated pods
+        for pod_key, po in self.changed_pods.items():
+            if po.apply_ns_policy():
+                if po.remove_default_profile():
+                    self.poRV = po.RV
+                    self.pods[pod_key] = po
+                    del self.changed_pods[pod_key]
+
+        # As endpoint lists change, create/add new profiles to pods as
+        # necessary
+        # TODO: if some object not ready for resync, finish loop w/out moving to
+        # processed list
+        for ep_key, ep in self.changed_endpoints.items():
+
+            # Get policy of the Service and namespace. Uses same namespace/name
+            # dict key.
+            try:
+                svc_policy = self.services[ep_key].type
+            except KeyError:
+                _log.warning("Service %s not yet in store" % ep_key)
+                continue
+
+            try:
+                ns_policy = self.namespaces[ep.namespace].policy
+            except KeyError, AttributeError:
```

Comment:
What throws the AttributeError?  That seems like a different error than the Namespace not being in the datastore yet.